client/src/main/java/io/pravega/client/segment/impl/AsyncSegmentInputStreamImpl.java

```diff
@@ -181,11 +180,17 @@ public boolean isClosed() {
             return getConnection()
                     .whenComplete((connection, ex) -> {
                         if (ex != null) {
-                            log.warn("Exception while establishing connection with Pravega " +
-                                    "node", ex);
+                            log.warn("Exception while establishing connection with Pravega node", ex);
                             closeConnection(new ConnectionFailedException(ex));
                         }
-                    }).thenCompose(c -> sendRequestOverConnection(request, c));
+                    }).thenCompose(c -> sendRequestOverConnection(request, c)
+                            .whenComplete((reply, ex) -> {
+                                if (ex instanceof ConnectionFailedException) {
```

Comment:
Shouldn't you unwrap this exception?

Reply:
Consider using `Futures.exceptionallyExpecting` which will handle all of this and make your code more compact.

Reply:
The future returned  `sendRequestOverConnecton` is used by the overall retry logic. Hence 
```
Futures.exceptionallyExpecting(future, t -> t instanceof ConnectionFailedException, value)
``` 
cannot be used. Do let me know if there is a better way to compact the code.
(the only other option is to invoke `closeConnection(...)` as part of the predicate)

Reply:
Ok. I missed the fact that this is not really returning a value but executing something.